File: frontend/src/app/services/reservation.service.ts
import { Injectable } from '@angular/core';

import { Globals } from '../global/globals';
import { Observable } from 'rxjs';
import { Reservation } from '../dtos/reservation';
import {HttpClient, HttpParams} from '@angular/common/http';


@Injectable({
  providedIn: 'root'
})
export class ReservationService {


  private reservationBaseUri: string = this.globals.backendUri + '/reservations';

  constructor(private httpClient: HttpClient, private globals: Globals) { }

  /**
   * Loads all reservations from the backend
   */
  public getAllReservations(): Observable<Reservation[]> {
    console.log('Get all reservations');
    return this.httpClient.get<Reservation[]>(this.reservationBaseUri + '/');
  }

  /**
   * Loads all reservations within the given time range from the backend
   * @param startDateTime start of the time range
   * @param endDateTime end of the time range
   */
  public getReservationsByStartAndEndDateTime(startDateTime: string, endDateTime: string): Observable<Reservation[]> {
    console.log('Get all reservations within the given time range from the backend');
    return this.httpClient.get<Reservation[]>(this.reservationBaseUri + '?startDateTime=' + startDateTime + '&endDateTime=' + endDateTime);
  }

    /**
   * Loads all reservations, except the reservation with the specified ID, within the given time range from the backend
   * @param startDateTime start of the time range
   * @param endDateTime end of the time range
   * @param idOfReservationToBeIgnored ID of the reservation to be ignored.
   */
  public findByStartAndEndDateTimeIgnoreReservationWithId(startDateTime: string, endDateTime: string, idOfReservationToBeIgnored: number): Observable<Reservation[]> {
    console.log('Get all reservations, except the reservation with the specified ID, within the given time range from the backend');
    return this.httpClient.get<Reservation[]>(this.reservationBaseUri + '?startDateTime=' + startDateTime + '&endDateTime=' + endDateTime + '&idOfReservationToBeIgnored=' + idOfReservationToBeIgnored );
  }

  /**
   * Creates a new reservation
   * @param reservation reservation to create
   */
  public createReservation(reservation: Reservation): Observable<Reservation> {
    console.log('ReservationService: Creating reservation.');
    return this.httpClient.post<Reservation>(this.reservationBaseUri + '/', reservation);
  }

  /**
   * Deletes the given reservation.
   * @param reservation the reservation to be deleted
   */

  public deleteReservation(reservation: Reservation): Observable<void> {
    console.log('ReservationService: Deleting reservation with id: ' + reservation.id);
    return this.httpClient.delete<void>(this.reservationBaseUri + '/' + reservation.id);
  }

/**
 * Loads the reservation with the specified id
 * @param id ID of the reservation to load
 */
  public getReservationWithId(id: number): Observable<Reservation> {
    console.log('getReservationWithId:' + id);
    return this.httpClient.get<Reservation>(this.reservationBaseUri + '/' + id);
  }


  /**
   * Updates the given reservation
   * @param reservation reservation with updated data
   */
  public updateReservation(reservation: Reservation): Observable<Reservation>  {
    console.log('updateReservation:' + JSON.stringify(reservation));
    return this.httpClient.put<Reservation>(this.reservationBaseUri, reservation);
  }

  /**
   * Filters saved reservations by the given parameters.
   * @param guestName     the name of the guest that made the reservation(s) to find.
   * @param startDateTime the earliest Date and Time of the reservation(s) to find. (exclusive)
   * @param endDateTime   the latest Date and Time of the reservations(s) to find. (exclusive)
   * @param tableNum      the table number of a table shall be part of the reservation(s) to find.
   */
  public filterReservations(guestName: string, startDateTime: string, endDateTime: string, tableNum: string): Observable<Reservation[]> {
    console.log('searching');
    guestName = !guestName ? '' : guestName;
    startDateTime = !startDateTime ? '' : startDateTime;
    endDateTime = !endDateTime ? '' : endDateTime;
    tableNum = !tableNum ? '' : tableNum;
    const httpParams = new HttpParams()
      .set('guestName', guestName)
      .set('startDateTime', startDateTime)
      .set('endDateTime', endDateTime)
      .set('tableNum', tableNum);
    const options = {params: httpParams};
    return this.httpClient.get<Reservation[]>(this.reservationBaseUri + '/filter', options);
  }

}

File: backend/src/main/java/com/spring/restaurant/backend/service/ReservationService.java
package com.spring.restaurant.backend.service;

import com.spring.restaurant.backend.entity.Reservation;
import com.spring.restaurant.backend.exception.CurrentlyUnavailableException;
import com.spring.restaurant.backend.exception.ValidationException;

import java.time.LocalDateTime;
import java.util.List;

public interface ReservationService {


    /**
     * Find all reservations by start date ascending.
     *
     * @return ordered list of all reservations (ascending by start date time)
     */
    List<Reservation> findAll();

    // TODO add NotFoundException
    /**
     * Find a single reservation by ID
     * @param id the ID of the reservation to find.
     * @return the reservation.
     */
    Reservation findOne(Long id);

    /**
     * Finds reservations in the time interval [startDateTime, endDateTime).
     * This means that the interval is open on side of endDateTime.
     * @param startDateTime the start date time of the time interval (inclusive)
     * @param endDateTime the end date of the of the time interval (exclusive)
     * @return reservations in the given time interval. Might be an empty list.
     * @throws ValidationException  will be thrown if the validation of the time range fails.
     * @throws CurrentlyUnavailableException will be thrown if something goes wrong during the data processing.
     */
    List<Reservation> findByStartAndEndDateTime(LocalDateTime startDateTime, LocalDateTime endDateTime);



    /**
     * Creates the given reservation.
     *
     * @param reservation the reservation to create.
     * @return the created reservation.
     * @throws ValidationException           will be thrown if the validation of the reservation fails.
     * @throws CurrentlyUnavailableException will be thrown if something goes wrong during the data processing.
     */
    Reservation createReservation(Reservation reservation);


    // TODO add NotFoundException
    /**
     * Deletes the reservation with the given id.
     * @param id the id of the reservation, which is to be deleted.
     * @throws CurrentlyUnavailableException will be thrown if something goes wrong during the data processing.
     */
    void deleteReservationById(Long id);

    /**
     * Filters saved reservations by the given parameters.
     * @param guestName     the name of the guest that made the reservation(s) to find.
     * @param startDateTime the earliest Date and Time of the reservation(s) to find. (exclusive)
     * @param endDateTime   the latest Date and Time of the reservations(s) to find. (exclusive)
     * @param tableNum      the table number of a table that shall be part of the reservation(s) to find.
     * @return a list of reservations that fit the the query (all, if every parameter is null)
     */
    List<Reservation> search(String guestName, String startDateTime, String endDateTime, Long tableNum);

    // TODO add NotFoundException
    /**
     * Modifies a reservation.
     *
     * @param reservation to modify
     * @return modified reservation
     */
    Reservation updateReservation(Reservation reservation);
}




